swift/trainers/trainers.py

```diff
@@ -179,24 +179,45 @@ def compute_loss(self, model, inputs, return_outputs=None):
         loss, outputs = super().compute_loss(model, inputs, True)
         preds = outputs.logits.argmax(dim=2)[..., :-1]
         labels = inputs['labels'][..., 1:]
-        masks = labels != -100
-        acc_strategy = getattr(self.args, 'acc_strategy', 'token')
-        acc: Tensor
-        if acc_strategy == 'sentence':
-            acc_list = []
-            for i, m in enumerate(masks):
-                acc_list.append(
-                    torch.all(preds[i, m] == labels[i,
-                                                    m]).to(torch.int64).item())
-            acc = torch.tensor(acc_list, device=preds.device).float().mean()
+        if not use_torchacc():
+            masks = labels != -100
+            acc_strategy = getattr(self.args, 'acc_strategy', 'token')
+            acc: Tensor
+            if acc_strategy == 'sentence':
+                acc_list = []
+                for i, m in enumerate(masks):
+                    acc_list.append(
+                        torch.all(preds[i, m] == labels[i, m]).to(
+                            torch.int64).item())
+                acc = torch.tensor(
+                    acc_list, device=preds.device).float().mean()
+            else:
+                acc = (preds[masks] == labels[masks]).float().mean()
+            if model.training:
+                if 'acc' not in self._custom_metrics:
+                    self._custom_metrics['acc'] = torch.tensor(0.).to(
+                        self.args.device)
+                self._custom_metrics[
+                    'acc'] += acc / self.args.gradient_accumulation_steps
         else:
-            acc = (preds[masks] == labels[masks]).float().mean()
-        if model.training:
-            if 'acc' not in self._custom_metrics:
-                self._custom_metrics['acc'] = torch.tensor(0.).to(
-                    self.args.device)
-            self._custom_metrics[
-                'acc'] += acc / self.args.gradient_accumulation_steps
+            acc_strategy = getattr(self.args, 'acc_strategy', 'token')
+            acc: Tensor
+            if acc_strategy == 'sentence':
+                acc_list = []
+                for i, m in enumerate(masks):
+                    acc_list.append(
+                        torch.all(preds[i, m] == labels[i, m]).to(
+                            torch.int64).item())
+                acc = torch.tensor(
+                    acc_list, device=preds.device).float().mean()
+            else:
+                acc = (preds == labels).float().mean()
+            if model.training:
```

Comment:
The code here has a bit high redundancy. Is it possible to optimize it?

Reply:
fixed